### PR TITLE
[GC] Add option G1BarrierSimple to use simple g1 post barrier

### PR DESCRIPTION
**Pull request by mmyxym**
Summary: Provide option G1BarrierSimple to use simple G1 post barrier for better mutator performance

Testing: CI pipeline

Reviewers: yude, yifeng

Issue: https://github.com/dragonwell-project/dragonwell21/issues/56

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1BarrierSet.cpp [hunk 1]
```diff
@@ -107,6 +107,12 @@ void G1BarrierSet::invalidate(JavaThread* thread, MemRegion mr) {
   }
   volatile CardValue* byte = _card_table->byte_for(mr.start());
   CardValue* last_byte = _card_table->byte_for(mr.last());
+  if (G1BarrierSimple) {
+    for (; byte <= last_byte; byte++) {
+      *byte = G1CardTable::dirty_card_val();
+    }
```
Comment (leveretconey) [inline src/hotspot/share/gc/g1/g1BarrierSet.cpp:113]: Why not use memset(), like what we did in dragonwell11?

File: src/hotspot/cpu/aarch64/gc/g1/g1BarrierSetAssembler_aarch64.cpp [hunk 2]
```diff
@@ -86,6 +86,28 @@ void G1BarrierSetAssembler::gen_write_ref_array_pre_barrier(MacroAssembler* masm
...
+
+    __ lea(end, Address(start, count, Address::lsl(LogBytesPerHeapOop))); // end = start + count << LogBytesPerHeapOop
+    __ sub(end, end, BytesPerHeapOop); // last element address to make inclusive
+    __ lsr(start, start, CardTable::card_shift());
+    __ lsr(end, end, CardTable::card_shift());
+    __ sub(count, end, start); // number of bytes to copy
+
+    __ load_byte_map_base(scratch);
+    __ add(start, start, scratch);
+    __ bind(L_loop);
```
Comment (linade) [inline src/hotspot/cpu/aarch64/gc/g1/g1BarrierSetAssembler_aarch64.cpp:103]: assert(dirty_card_val == 0)

File: src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp [hunk 3]
```diff
@@ -98,6 +100,36 @@ void G1BarrierSetAssembler::gen_write_ref_array_pre_barrier(MacroAssembler* masm
...
+    __ leaq(end, Address(addr, count, TIMES_OOP, 0));  // end == addr+count*oop_size
+    __ subptr(end, BytesPerHeapOop); // end - 1 to make inclusive
+    __ shrptr(addr, CardTable::card_shift());
+    __ shrptr(end, CardTable::card_shift());
+    __ subptr(end, addr); // end --> cards count
+
+    __ mov64(tmp, disp);
+    __ addptr(addr, tmp);
+    __ bind(L_loop);
+    __ movb(Address(addr, count, Address::times_1), 0);
```
Comment (linade) [inline src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp:125]: Use G1CardTable::dirty_card_val() like the rest of the code

File: src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp [hunk 4]
```diff
@@ -98,6 +100,36 @@ void G1BarrierSetAssembler::gen_write_ref_array_pre_barrier(MacroAssembler* masm
 
 void G1BarrierSetAssembler::gen_write_ref_array_post_barrier(MacroAssembler* masm, DecoratorSet decorators,
                                                              Register addr, Register count, Register tmp) {
+#ifdef _LP64
```
Comment (linade) [inline src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp:103]: G1BarrierSimple seems to permit X86. What happens on that platform?

**Comment by linade** [comment]
Probably not enough functional tests which will concern our coverage goals..

File: src/hotspot/share/gc/g1/g1BarrierSet.cpp (hunk 1 above)
Comment (mmyxym) [reply src/hotspot/share/gc/g1/g1BarrierSet.cpp]: fixed

File: src/hotspot/cpu/aarch64/gc/g1/g1BarrierSetAssembler_aarch64.cpp (hunk 2 above)
Comment (mmyxym) [reply src/hotspot/cpu/aarch64/gc/g1/g1BarrierSetAssembler_aarch64.cpp:103]: fixed

File: src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp (hunk 3 above)
Comment (mmyxym) [reply src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp]: fixed

File: src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp (hunk 4 above)
Comment (mmyxym) [reply src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp:103]: The option will be ignored.

File: src/hotspot/share/gc/g1/g1Arguments.cpp [hunk 5]
```diff
@@ -177,6 +177,25 @@ void G1Arguments::initialize() {
     FLAG_SET_ERGO(ParallelGCThreads, 1);
   }
 
+  if (G1BarrierSimple) {
+#if !defined(_LP64) || !(defined(X86) || defined(AARCH64))
```
Comment (linade) [inline src/hotspot/share/gc/g1/g1Arguments.cpp:181]: X86 is allowed here?

Comment (mmyxym) [reply src/hotspot/share/gc/g1/g1Arguments.cpp:181]: only x86 32bit is not supported.

Comment (linade) [reply src/hotspot/share/gc/g1/g1Arguments.cpp:181]: ok, took me a while to see this means !(x86_64 || aarch64)

**Review by linade — APPROVED** [review]

**Review by leveretconey — APPROVED** [review]

**Pull request merged**